File: 39.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Integer right triangles". The
problem is: For which value of p <= 1000, is the number of solutions maximised?
"""

from __future__ import division
from collections import defaultdict
import math


def main():
    """Integer right triangles"""

    # Constants
    MAX_SIDE = int(1000 / 2 + 1)

    count = defaultdict(int)

    for side1 in range(1, MAX_SIDE):
        for side2 in range(side1, MAX_SIDE):
            side3 = math.sqrt(side1 * side1 + side2 * side2)
            if int(side3) == side3:
                count[side1 + side2 + int(side3)] += 1

    print(max(count, key=count.get))

if __name__ == '__main__':
    main()

File: 39.js
#!/usr/bin/env node

// Integer right triangles

var MAX_SIDE = 1000 / 2;

var count = [];

for (var side1 = 1; side1 <= MAX_SIDE; side1++) {
    for (var side2 = side1; side2 <= MAX_SIDE; side2++) {
        var side3 = Math.sqrt(side1 * side1 + side2 * side2);
        if (Math.floor(side3) == side3) {
            if (!count[side1 + side2 + side3]) {
                count[side1 + side2 + side3] = 0;
            }
            count[side1 + side2 + side3]++;
        }
    }
}

var idx = 0;
var max = count[idx];
for (var i = 1; i < count.length; i++) {
    if (!max || count[i] > max) {
        idx = i;
        max = count[i];
    }
}

console.log(idx);
